### PR TITLE
daml-lf: rename Map to TextMap in archive proto

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This PR advance the state of Generic Map #2256.

In order to avoid confusion with Generic Map, we rename `Map` to `TextMap` in:
- archive protobuf defitions
- Haskell AST
- Scala AST

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [X] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags, if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs [hunk 1]
```diff
@@ -91,7 +91,7 @@ serializabilityConditionsType world0 _version mbModNameTpls vars = go
         BTBool -> noConditions
         BTList -> Left URList  -- 'List' is used as a higher-kinded type constructor.
         BTOptional -> Left UROptional  -- 'Optional' is used as a higher-kinded type constructor.
-        BTMap -> Left URMap  -- 'Map' is used as a higher-kinded type constructor.
+        BTTextMap -> Left URMap  -- 'Map' is used as a higher-kinded type constructor.
```
Comment (associahedron) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Serializability.hs]:
```suggestion
        BTTextMap -> Left URMap  -- 'TextMap' is used as a higher-kinded type constructor.
```

**Comment by remyhaemmerle-da** [comment]
> Doesn't this constitute a breaking change at the DALF level? I don't think that's a good idea.

At the archive level, it is not a breaking change because 
- It does not change the serialization.
- We provide all versions of protobuf definition where the name is not change. 

  * ``daml-lf/archive/src/main/protobuf/com/digitalasset/daml_lf_1_6/daml_lf_1.proto``
  * ``daml-lf/archive/src/main/protobuf/com/digitalasset/daml_lf_1_7/daml_lf_1.proto``

When we will introduce a new version of daml-lf we will freeze a new Protobuf definition.
If somebody wants to support this new version of LF, it will have to use this new definition and will have to change its code anyway.

**Comment by cocreature** [comment]
Renaming it in the archive proto while keeping the old name in the Ledger API value proto seems more confusing to me than the status quo. Are we also planning to rename it there? Otherwise I would prefer to stick to the current naming.

**Comment by remyhaemmerle-da** [comment]
@cocreature wrote:
> Renaming it in the archive proto while keeping the old name in the Ledger API value proto seems more confusing to me than the status quo. Are we also planning to rename it there? Otherwise I would prefer to stick to the current naming.

I plan to rename everywhere I can. I will have to  negotiate hardly with @gerolf-da  to change it in the ledger API level.
Anyway 
- the confusion is already there since the type is called `TextMap` in Daml and `Map` in DAML-LF.
  I think is is simpler to change it everywhere we can and clearly comment that `TextMap` is called is called `Map` in the ledger API for backward compatibility reasons.
- We have to do think gradually.

**Comment by remyhaemmerle-da** [comment]
If necessary I will create version of the protobuf definition of the ledger API as I have done for the archive proto.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs [hunk 2]
```diff
@@ -215,8 +215,8 @@ data BuiltinExpr
   | BEGreater    !BuiltinType    -- :: t -> t -> Bool, where t is the builtin type
   | BEToText     !BuiltinType    -- :: t -> Text, where t is one of the builtin types
                                  -- {Int64, Decimal, Text, Timestamp, Date, Party}
-
-  -- Decimal arithmetic
+                                 
+  -- Decimal arithmetic
```
Comment (associahedron) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs]:
```suggestion
  -- Decimal arithmetic
```

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -3052,14 +3054,17 @@ The deserialization process will reject any DAML-LF 1.2 (or earlier)
...
 [*Available in versions >= 1.3*]
 
 The deserialization process will reject any DAML-LF 1.2 (or earlier)
-program using the builtin functions : ``MAP_EMPTY``, ``MAP_INSERT``,
-``MAP_LOOKUP``, ``MAP_DELETE``, ``MAP_LIST``, ``MAP_SIZE``,
+program using the builtin functions : ``TEXTMAP_EMPTY``,
+``TEXTMAP_INSERT``, ``TEXTMAP_LOOKUP``, ``TEXTMAP_DELETE``,
+``TEXTMAP_LIST``, ``TEXTMAP_SIZE``,
+
+TextMpa was called ``Map`` in versions <= 1.dev.
```
Comment (associahedron) [inline daml-lf/spec/daml-lf-1.rst]:
```suggestion
``'TextMap'`` was called ``'Map'`` in versions <= 1.dev.
```

**Comment by gerolf-da** [comment]
The problem of inconsistent naming already exists today with `TextMap` in DAML and `Map` in the Ledger API. Considering that way fewer users will actually use DAML LF, I'd say we can change `Map` -> `TextMap` in DAML LF to have better and consistent naming there.

**Comment by cocreature** [comment]
DAML and DAML-LF are inconsistent in a number of ways (e.g. even ADTs are defined differently in DAML where we have sums of products and DAML-LF where we have separate variants and records). Personally, I care more about being consistent across the Ledger API and DAML-LF since the DAML-LF defines what I will get over the Ledger API. That said, I don’t feel too strongly about this so I’m fine with this change.

**Pull request merged**